File: frontend/src/types/billing.ts
export interface SubscriptionPlan {
  id: string;
  name: string;
  slug: string;
  description: string | null;
  price_monthly: string;
  price_yearly: string;
  currency: string;
  max_users: number | null;
  max_storage_gb: number | null;
  included_modules: string | null;
  stripe_product_id: string | null;
  stripe_monthly_price_id: string | null;
  stripe_yearly_price_id: string | null;
  is_active: boolean;
  sort_order: number;
  created_at: string;
  updated_at: string;
}

export interface SubscriptionPlanCreate {
  name: string;
  slug: string;
  description?: string;
  price_monthly?: string;
  price_yearly?: string;
  currency?: string;
  max_users?: number;
  max_storage_gb?: number;
  included_modules?: string;
  sort_order?: number;
}

export interface SubscriptionPlanUpdate {
  name?: string;
  description?: string;
  price_monthly?: string;
  price_yearly?: string;
  max_users?: number;
  max_storage_gb?: number;
  included_modules?: string;
  is_active?: boolean;
  sort_order?: number;
}

export interface Subscription {
  id: string;
  company_id: string;
  plan_id: string;
  status: "trialing" | "active" | "past_due" | "canceled" | "unpaid";
  billing_interval: "monthly" | "yearly";
  stripe_customer_id: string | null;
  stripe_subscription_id: string | null;
  current_period_start: string | null;
  current_period_end: string | null;
  trial_end: string | null;
  canceled_at: string | null;
  current_user_count: number;
  current_storage_mb: number;
  created_at: string;
  updated_at: string;
  plan: SubscriptionPlan | null;
  company_name: string | null;
}

export interface SubscriptionCreate {
  company_id: string;
  plan_id: string;
  billing_interval?: string;
}

export interface PaginatedSubscriptions {
  items: Subscription[];
  total: number;
  page: number;
  per_page: number;
}

export interface BillingEvent {
  id: string;
  company_id: string;
  subscription_id: string | null;
  event_type: string;
  amount: string | null;
  currency: string;
  stripe_event_id: string | null;
  stripe_invoice_id: string | null;
  metadata_json: string | null;
  created_at: string;
}

export interface PaginatedBillingEvents {
  items: BillingEvent[];
  total: number;
  page: number;
  per_page: number;
}

export interface BillingStats {
  total_subscriptions: number;
  active_subscriptions: number;
  past_due: number;
  canceled: number;
  mrr: string;
  total_revenue_30d: string;
}

File: backend/app/schemas/billing.py
from datetime import datetime
from decimal import Decimal

from pydantic import BaseModel, Field


# ---------------------------------------------------------------------------
# Subscription Plans
# ---------------------------------------------------------------------------


class SubscriptionPlanCreate(BaseModel):
    name: str
    slug: str
    description: str | None = None
    price_monthly: Decimal = Decimal("0.00")
    price_yearly: Decimal = Decimal("0.00")
    currency: str = "USD"
    max_users: int | None = None
    max_storage_gb: int | None = None
    included_modules: str | None = None  # JSON string
    stripe_product_id: str | None = None
    stripe_monthly_price_id: str | None = None
    stripe_yearly_price_id: str | None = None
    sort_order: int = 0


class SubscriptionPlanUpdate(BaseModel):
    name: str | None = None
    description: str | None = None
    price_monthly: Decimal | None = None
    price_yearly: Decimal | None = None
    max_users: int | None = None
    max_storage_gb: int | None = None
    included_modules: str | None = None
    stripe_product_id: str | None = None
    stripe_monthly_price_id: str | None = None
    stripe_yearly_price_id: str | None = None
    is_active: bool | None = None
    sort_order: int | None = None


class SubscriptionPlanResponse(BaseModel):
    id: str
    name: str
    slug: str
    description: str | None = None
    price_monthly: Decimal
    price_yearly: Decimal
    currency: str
    max_users: int | None = None
    max_storage_gb: int | None = None
    included_modules: str | None = None
    stripe_product_id: str | None = None
    stripe_monthly_price_id: str | None = None
    stripe_yearly_price_id: str | None = None
    is_active: bool
    sort_order: int
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


# ---------------------------------------------------------------------------
# Subscriptions
# ---------------------------------------------------------------------------


class SubscriptionCreate(BaseModel):
    company_id: str
    plan_id: str
    billing_interval: str = "monthly"
    stripe_customer_id: str | None = None


class SubscriptionResponse(BaseModel):
    id: str
    company_id: str
    plan_id: str
    status: str
    billing_interval: str
    stripe_customer_id: str | None = None
    stripe_subscription_id: str | None = None
    current_period_start: datetime | None = None
    current_period_end: datetime | None = None
    trial_end: datetime | None = None
    canceled_at: datetime | None = None
    current_user_count: int
    current_storage_mb: int
    created_at: datetime
    updated_at: datetime
    plan: SubscriptionPlanResponse | None = None
    company_name: str | None = None

    class Config:
        from_attributes = True


class ChangePlanRequest(BaseModel):
    plan_id: str
    billing_interval: str | None = None


class PaginatedSubscriptions(BaseModel):
    items: list[SubscriptionResponse]
    total: int
    page: int
    per_page: int


# ---------------------------------------------------------------------------
# Billing Events
# ---------------------------------------------------------------------------


class BillingEventResponse(BaseModel):
    id: str
    company_id: str
    subscription_id: str | None = None
    event_type: str
    amount: Decimal | None = None
    currency: str
    stripe_event_id: str | None = None
    stripe_invoice_id: str | None = None
    metadata_json: str | None = None
    created_at: datetime

    class Config:
        from_attributes = True


class PaginatedBillingEvents(BaseModel):
    items: list[BillingEventResponse]
    total: int
    page: int
    per_page: int


# ---------------------------------------------------------------------------
# Billing Stats
# ---------------------------------------------------------------------------


class BillingStats(BaseModel):
    total_subscriptions: int
    active_subscriptions: int
    past_due: int
    canceled: int
    mrr: Decimal  # Monthly Recurring Revenue
    total_revenue_30d: Decimal
